### PR TITLE
fix(test): improving unit tests results in IE11 and Firefox 38

### PR DESCRIPTION
**Pull request by marclaval**
A first batch of fixes to improve unit tests results in IE11 and Firefox 38.
With them, the numbers of failures is reduced as follows:
- from 124 to 30 in IE11
- from 39 to 23 in Firefox 38

More fixes will follow, aiming to reach 0 in both.

### CODE REVIEW

File: modules/angular2/src/dom/browser_adapter.dart [hunk 1]
```diff
@@ -323,4 +323,7 @@ class BrowserDomAdapter extends GenericBrowserDomAdapter {
     var baseUri = Uri.parse(uri);
     return baseUri.path;
   }
+  getUserAgent() {
```
Comment (vicb) [inline modules/angular2/src/dom/browser_adapter.dart]: return type

File: modules/angular2/test/change_detection/pipes/promise_pipe_spec.js [hunk 2]
```diff
@@ -12,6 +13,7 @@ export function main() {
     var pipe;
     var completer;
     var ref;
+    var timer = (DOM.getUserAgent().indexOf("Trident") > -1) ? 50 : 0;
```
Comment (vicb) [inline modules/angular2/test/change_detection/pipes/promise_pipe_spec.js]: add a comment

File: modules/angular2/test/core/zone/ng_zone_spec.js [hunk 3]
```diff
@@ -16,12 +16,14 @@ import {
...
 import {BaseException} from 'angular2/src/facade/lang';
+import {DOM} from 'angular2/src/dom/dom_adapter';
 
 import {NgZone} from 'angular2/src/core/zone/ng_zone';
 
+var isIE = DOM.getUserAgent().indexOf("Trident") > -1;
 // Schedules a macrotask (using a timer)
-function macroTask(fn: Function): void {
-  _zone.runOutsideAngular(() => TimerWrapper.setTimeout(fn, 1));
+function macroTask(fn: Function, timer = 1): void {
```
Comment (vicb) [inline modules/angular2/test/core/zone/ng_zone_spec.js]: add a comment

**Comment by vicb** [comment]
Are tests executed on Sauce for those browsers ?

**Comment by mlaval** [comment]
Comments integrated.

 About Sauce, it is not yet used, but it would be great to set it up once some browsers are green.

File: karma-js.conf.js [hunk 4]
```diff
@@ -24,7 +24,8 @@ module.exports = function(config) {
       'node_modules/reflect-metadata/Reflect.js',
       'tools/build/file2modulename.js',
       'test-main.js',
-      {pattern: 'modules/**/test/**/static_assets/**', included: false, watched: false}
+      {pattern: 'modules/**/test/**/static_assets/**', included: false, watched: false},
+      'modules/angular2/src/test_lib/shims_for_IE.es6'
```
Comment (vicb) [inline karma-js.conf.js]: should it be .js ?

**Comment by vicb** [comment]
LGTM after you add a missing comment to explain why we need one hack

**Pull request closed without merging**